### DISCUSSION TITLE
Should Hex Fiend switch to Swift?

### DISCUSSION THREAD

**Discussion by kainjow** [HexFiend/HexFiend#373] [Ideas]
HF has been an Objective-C/AppKit app all its life. Swift is mature now and Apple has made it clear it’s the future. It’s certainly where all feature development is going. SwiftUI also makes many previous UI tasks easy (but also making others difficult).

I think moving to Swift would be good. Some things like working with low level bytes is less natural, but with it comes greater safety. The move can be done file by file. I was thinking of starting with some tests, then app classes.

Regarding the framework which is meant to be used by other apps, ideally not keeping Objective-C compatibility would be best. It could start with compatibility until all classes have been ported. Maybe it becomes a Swift package, but mixing Objective-C and Swift is not possible based on my understanding.

**Comment by zorgiepoo** [comment]
I use the HexFiend framework in a 10-year old obj-c app that I update once/twice a year so I will probably keep using an obj-c version for a long time (I update/recompile HexFiend on a lesser cadence). Unfortunately I also subclass several HexFiend classes, essentially like a fork, like to [customize not starting at 0x0](https://github.com/zorgiepoo/Bit-Slicer/blob/main/Bit%20Slicer/ZGLineCountingView.m) or pull in the [DataInspector editor](https://github.com/zorgiepoo/Bit-Slicer/tree/main/deps/HexFiend) that isn't part of the framework (I think; sorry, never mentioned before :P).

If you use any Swift code I would probably raise the minimum deployment target to 10.14.6, to avoid including extra Swift baggage that doesn't need to be bundled in your app. As a heads up, Xcode 14 is probably [going to raise the minimum possible supported version to 10.13](https://developer.apple.com/documentation/Xcode-Release-Notes/xcode-14-release-notes) anyway though.

> Maybe it becomes a Swift package, but mixing Objective-C and Swift is not possible based on my understanding.

Not sure if this is true for binary packages. Pretty sure it's true for source/regular ones, though.

Maybe I'm biased but I don't personally see Swift as easier/better when it comes to raw memory manipulation, and outside of like a Preferences window I don't see SwiftUI being better suited (HexFiend has very custom controls / AppKit intricacy if I'm remembering correctly.. maybe I'm wrong). IMO, best place to first start using Swift may be like high-level UI components specific to the app and outside the framework, and any testing. code if there's any.

Thanks for the framework btw :)

**Comment by kainjow** [comment]
Thanks for the response. Good to know where Hex Fiend is being used as well!

Not starting at 0 sounds like some feature requests that have come in for essentially viewing a slice. And having the data inspector classes in the framework I see could be useful, but those are some of the classes I see being targeted for moving to Swift too. Maybe data inspector could become its own Swift package. I also think the binary templates code could become its own library and have the HF hooks abstracted.

For raw memory manipulation, I agree with Swift being more difficult, but I feel that's not a big part of the codebase, so once some standards are setup it seems like it should be fine. Also figuring out all those Swift classes that do lower-level pointer manipulation - I always have to look up the docs since they're less intuitive (at least to me).

I think SwiftUI could be helpful, for example with simplifying the data inspector UI, or a rewrite of the Find controls. I'm not sure about the main scrollable view, as I haven't looked to see yet if Canvas can be scrolled. Maybe it's still an area they are investigating.

I also did see the newer minimum target requirements with Xcode 14. Hopefully they make that an error in the release build if you target an older OS.

Thanks again for your feedback.

**Reply by zorgiepoo** [reply]
> Not starting at 0 sounds like some feature requests that have come in for essentially viewing a slice

In my case I feed HexFiend with data that I obtain in real-time (from a process' virtual memory) and not from a concrete file, and I decide the subset range I want to show.

I had forgotten, but I also `extern "C"` link `boyer_moore_helper()` 😉 and took HexFiend's code [for preparing it](https://github.com/zorgiepoo/Bit-Slicer/blob/main/Bit%20Slicer/HFByteArray_FindReplace.cpp#L27). Not for hex editing purposes, but because my app has search functionality elsewhere too. I suppose that is [one such example](https://ridiculousfish.com/blog/posts/old-age-and-treachery.html) where performance may need to be carefully evaluated.

> I'm not sure about the main scrollable view, as I haven't looked to see yet if Canvas can be scrolled. Maybe it's still an area they are investigating.

HexFiend uses NSScroller and doesn't use NSScrollView as well for the main view right? Not sure if that has to do with performance.

**Comment by kainjow** [comment]
Yes it uses custom scroll behavior, so you don't get that rubber band bouncing effect sadly.

I wonder if with SwiftUI a LazyVStack could be used for HF. Not sure if selection could be made to work though.

I'll be watching ExtensionFoundation for macOS 13. Looks like HF could possibly support Swift binary template plugins in the future..

**Comment by rtharston** [comment]
Well, I am glad I found this discussion! I created a couple of PRs earlier this week with a couple classes converted to Swift, just because I thought it would be an interesting experience to migrate an app I use to Swift. Then when I forked the repo I saw there already was a branch that started migrating a few things to Swift, so I decided to make PRs for my Swift code instead of just keeping it in my own fork.

Now that I found this discussion the presence of that branch makes a bit more sense. I'd like to through in some of my own thoughts.

First, I have no production experience with Obj-C. I have tried to use it a bit a few times in the past, but by the time I started working in macOS development professionally Swift had already taken over new development. That means I can read most of Obj-C, but I'm no expert, and contributing to codebases like HexFiend's would be much more difficult for me to do in Obj-C than Swift.

(As an aside, I have plenty of experience with C and C++, so it is just the Objective part of Objective-C that I struggle a bit with.)

So, part of the decision to stay on Obj-C vs migrating to Swift also depends on what talent you might want to attract to help work on the code (if that is a point of concern for you). Old talent will likely be in Obj-C, new talent will more likely know Swift. Not sure how much that matters to you, but I figured it was worth pointing out.

Low level access.
Yes, Obj-C (through C) has simple low level access, but that is also what makes it really hard to get right, and really easy to get really wrong, unless you have lots of experience. In my experience, Swift is a bit harder to use for low level tasks, but that is _on purpose_ to help you think properly about what you are doing and guide you to doing it the right way. And as you say, that is a small part of the app, so I wouldn't worry too much about that.
And you can always just leave the part in Obj-C (or even just C) and call into it from Swift if you want! My migration strategy would be to move view code over first and do the low level stuff last, since it is already working and doesn't _need_ to be replaced.

Speaking of view code.
Why are you already bringing up SwiftUI? I love it. We are using it heavily now at the company I work for, but only because we recently moved our minimum target to macOS 11 Big Sur. SwiftUI requires at least 10.15 Catalina, but the general consensus seems to be that it isn't good enough to do much more than very simple views like preference panes. It has gotten much better from year to year. Seeing how HexFiend 2's current support is all the way back to 10.11 I can't see you wanting to make that big of a jump right now. You _can_, and I won't stop you (I'd love to help move some of the views to SwiftUI actually!), but I don't know if that is prudent yet.
As for the minimum target moving to 10.13 in Xcode 14, I am using Xcode 13 for my fork so I wouldn't move that up, but that requires staying in macOS 12 Monterey. Xcode 13 isn't supported in macOS 14 Ventura. If you want to make the call on moving to Xcode 14 let me know, since that brings Swift 5.7 which brings some nice new features. (Might not be worth talking about async await here, since a) I don't know how much it would be useful in HexFiend yet, and b) it requires a minimum target of macOS 10.15).

You can very easily just translate all your existing Obj-C code to Swift and keep all your architecture the exact same as it is today. One might argue why you would want to go through the trouble if you are leaving things the same, but personally I find the advantages of Swift worth it. And once you do feel comfortable with making the jump to a much newer version of macOS it will let you easily start moving views to SwiftUI (which you can actually also do with most of your code in Obj-C, it is just much easier if the other code is Swift).

In short, don't forget that Swift was very carefully designed to coexist with Obj-C, so the transition can be piecemeal, as it makes sense. Some Swift classes will need to be labeled with @objc, but over time you can start taking that label off and seeing more of the benefits of Swift.

And I know you can put both Swift and Obj-C code in the same Swift Package because most of the packages our products uses do that. They are packages that have been around for a long time (starting as Cocoapod packages) and still are mostly Obj-C. One example is [Firebase](https://github.com/firebase/firebase-ios-sdk).

**Comment by kainjow** [comment]
d23783fe16084e03a9b2d3fd313d7e6ca2dcdcfe is the first commit to use Swift. I'm thinking all new classes will be Swift only. Will probably stick to Objective-C for most existing code changes for now.

**Comment by andrews05** [comment]
Great to hear some movement towards Swift has started! I would personally be thrilled if the framework could be separated from the app and made available via Swift Package Manager. That's probably a big ask though.

**Reply by rtharston** [reply]
I'm excited to see movement too! I've reopened my first PR with a small Swift change: #433
(I kind of regret not leaving it open earlier, maybe my code could have been the first Swift in the repo. 😅)

I agree that it is simpler and safer to start with adding Swift in new class first. Migrating Obj-C code isn't terribly hard, but that is tried and tested code. (That won't stop me from trying of course, so I hope you can forgive the occasional small migration PR. 😁)

**Comment by kainjow** [comment]
Moving to Swift Package Manager I think would require creating two packages. One for Objective-C sources, and one for Swift. I'd have to look into it, and what else the current framework depends upon.

I'm also fine moving small/simpler files to Swift, and anything that will have upcoming larger changes. I think this would be a good time to expand unit tests, for things that can be easily unit testable, by *first* adding the tests (in Swift, of course), then porting the code over.

**Comment by JetForMe** [comment]
FWIW, I do think you should move to Swift, but not SwiftUI.

**Comment by kainjow** [comment]
SwiftUI seems great for basic control layout. Which is the majority of HF except its scrollable views. The problem there I believe with SwiftUI is it’s want to retain what was rendered, vs AppKit views like NSCollectionView or NSTableView, or even NSView’s ability to render any rect with a custom drawing context. SwiftUI’s Canvas seems close but I think is still lacking scroll capabilities. Also macOS seems to be their lowest priority with it vs other platforms. I want to try using it for the preferences window as a starting point.

**Reply by rtharston** [reply]
SwiftUI support on macOS requires 10.15 and above, but you only get very basic views in 10.15, with more views added each year. That means you'd need to bump the minimum version to 10.15.

But the worst part is that each version of macOS doesn't add new views, it changes and breaks things. For example, I believe it is in macOS 11 that SwiftUI switched the direction of the y coordinate, so if you use a set position or offset it will be correct in one version and wrong in the other. (I can't remember which way it switched, just that we had to add code to handle it.)

And while using SwiftUI and NSViews together does work, they don't always get along if you aren't careful. The issues I saw most were around SwiftUI frame sizing and NSView auto layout. We needed to make a new settings pane, so I made it using SwiftUI and it worked great, except it wouldn't resize properly, and `padding` wasn't always respected. One of my coworkers eventually figured out the right magic incantation to get it to work, but it was a pain at first. (I wish I remembered what the solution was, sorry.)

So yeah, if you want to use it start small, so you can tackle issues that come up one at a time.
